### PR TITLE
tikv/txn: implement txn scheduler 

### PR DESCRIPTION
**Pull request by AndreMouche**
Hi,
    This PR implement scheduler for transactions:
    It will check the local latches first before committing a retryable transaction.

@tiancaiamao @coocood @zhangjinpeng1987 @disksing PTAL

### CODE REVIEW

**Comment by coocood** [comment]
@AndreMouche 
We need a configuration to enable it and disable it by default.

**Comment by ngaut** [comment]
Do we have any benchmark?

**Comment by AndreMouche** [comment]
yes, it's on our internal jira,  and we will retest before the merge. @ngaut

File: store/tikv/txn_scheduler.go [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+			txns:    nil,
+		}
+	}
+	return &txnScheduler{
+		latches: latch.NewLatches(txnSchedulerLatchesSize),
+		txns:    make([]txnMap, txnSchedulerMapsCount),
+	}
+}
+
+func (store *txnScheduler) execute(ctx context.Context, txn *twoPhaseCommitter, connID uint64) error {
```
Comment (coocood) [inline store/tikv/txn_scheduler.go]: s/store/scheduler

File: store/tikv/txn_scheduler.go [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+}
+
+func newTxnCommiter(ctx context.Context, committer *twoPhaseCommitter) *txnCommitter {
+	return &txnCommitter{
+		committer: committer,
+		ctx:       ctx,
+	}
+}
+
+func (txn *txnCommitter) commit(timeout bool) uint64 {
```
Comment (coocood) [inline store/tikv/txn_scheduler.go]: s/txn/tc

File: store/tikv/txn_scheduler.go [hunk 3]
```diff
@@ -0,0 +1,188 @@
...
+	"github.com/juju/errors"
+	"github.com/pingcap/tidb/sessionctx"
+	"github.com/pingcap/tidb/store/tikv/latch"
+	binlog "github.com/pingcap/tipb/go-binlog"
+	log "github.com/sirupsen/logrus"
+	"golang.org/x/net/context"
+)
+
+type txnCommitter struct {
+	committer *twoPhaseCommitter
```
Comment (coocood) [inline store/tikv/txn_scheduler.go]: s/committer/twoPC

File: store/tikv/txn_scheduler.go [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+func (t *txnMap) delete(startTS uint64) {
+	t.Lock()
+	defer t.Unlock()
+	delete(t.txns, startTS)
+}
+
+// txnScheduler is the scheduler to commit transactions.
+type txnScheduler struct {
+	latches *latch.Latches
+	txns    []txnMap
```
Comment (coocood) [inline store/tikv/txn_scheduler.go]: s/txns/txnMaps

File: store/tikv/txn_scheduler.go [hunk 5]
```diff
@@ -0,0 +1,188 @@
...
+	ch := make(chan error, 1)
+	newCommitter.ch = ch
+	store.putTxn(txn.startTS, newCommitter)
+	go store.run(txn.startTS)
+
+	err := errors.Trace(<-ch)
+	return errors.Trace(err)
+}
+
+func (store *txnScheduler) getMap(startTS uint64) *txnMap {
```
Comment (coocood) [inline store/tikv/txn_scheduler.go]: No need to use pointer.

File: store/tikv/txn_scheduler.go [hunk 6]
```diff
@@ -0,0 +1,188 @@
...
+	// for transactions which is not retryable, commit directly.
+	if !sessionctx.GetRetryAble(ctx) {
+		err := store.runForUnRetryAble(newCommitter)
+		return errors.Trace(err)
+	}
+	// for transactions which need to acquire latches
+	ch := make(chan error, 1)
+	newCommitter.ch = ch
+	store.putTxn(txn.startTS, newCommitter)
+	go store.run(txn.startTS)
```
Comment (coocood) [inline store/tikv/txn_scheduler.go]: We can call `run` directly without start another goroutine.

File: store/tikv/txn_scheduler.go (hunk 2 above)
Comment (coocood) [reply store/tikv/txn_scheduler.go]: s/timeout/stale

File: store/tikv/txn_scheduler.go [hunk 7]
```diff
@@ -0,0 +1,188 @@
...
+func newTxnCommiter(ctx context.Context, committer *twoPhaseCommitter) *txnCommitter {
+	return &txnCommitter{
+		committer: committer,
+		ctx:       ctx,
+	}
+}
+
+func (txn *txnCommitter) commit(timeout bool) uint64 {
+	if timeout {
+		err := errors.Errorf("startTS %d timeout", txn.committer.startTS)
```
Comment (coocood) [inline store/tikv/txn_scheduler.go]: `transaction startTS %d is stale`

File: store/tikv/txn_scheduler.go [hunk 8]
```diff
@@ -0,0 +1,188 @@
...
+	return errors.Trace(err)
+}
+
+// run a normal transaction, it needs to acquire latches before 2pc.
+func (store *txnScheduler) run(startTS uint64) {
+	txn := store.getTxn(startTS)
+	if txn == nil {
+		panic(startTS)
+	}
+	acquired, timeout := store.latches.Acquire(txn.lock)
```
Comment (coocood) [inline store/tikv/txn_scheduler.go]: s/timeout/stale

File: store/tikv/txn_scheduler.go [hunk 9]
```diff
@@ -0,0 +1,188 @@
...
+}
+
+func (t *txnMap) put(startTS uint64, txn *txnCommitter) {
+	t.Lock()
+	defer t.Unlock()
+	if t.txns == nil {
+		t.txns = make(map[uint64]*txnCommitter)
+	}
+	if _, ok := t.txns[startTS]; ok {
+		panic(fmt.Sprintf("The startTS %d shouldn't be used in two transaction", txn.committer.startTS))
```
Comment (coocood) [inline store/tikv/txn_scheduler.go]: Don't panic here, it will crash the tidb-server.

File: store/tikv/txn_scheduler.go [hunk 10]
```diff
@@ -0,0 +1,188 @@
...
+	}
+	log.Debug(txn.committer.connID, " finish txn with startTS:", txn.committer.startTS, " commitTS:", commitTS, " error:", err)
+	return errors.Trace(err)
+}
+
+// run a normal transaction, it needs to acquire latches before 2pc.
+func (store *txnScheduler) run(startTS uint64) {
+	txn := store.getTxn(startTS)
+	if txn == nil {
+		panic(startTS)
```
Comment (coocood) [inline store/tikv/txn_scheduler.go]: Don't panic here, it will crash the tidb-server.

File: store/tikv/txn_scheduler.go (hunk 9 above)
Comment (AndreMouche) [reply store/tikv/txn_scheduler.go]: I think it's a disaster if two or more transactions share the same `startTS`, I think panic should be ok here.

File: store/tikv/txn_scheduler.go [hunk 11]
```diff
@@ -0,0 +1,189 @@
...
+
+func (scheduler *txnScheduler) putTxn(startTS uint64, txn *txnCommitter) {
+	scheduler.getMap(startTS).put(startTS, txn)
+}
+
+// run the transaction directly which is not retryable.
+func (scheduler *txnScheduler) runForUnRetryAble(txn *txnCommitter) error {
+	commitTS, err := txn.execute()
+	if err == nil {
+		go scheduler.latches.RefreshCommitTS(txn.lock.RequiredSlots(), commitTS)
```
Comment (coocood) [inline store/tikv/txn_scheduler.go]: We don't need to start a new goroutine to do it.

File: store/tikv/txn_scheduler.go [hunk 12]
```diff
@@ -0,0 +1,189 @@
...
+	acquired, stale := scheduler.latches.Acquire(txn.lock)
+	if !stale && !acquired {
+		// wait for next wakeup
+		return
+	}
+	commitTS := txn.commit(stale)
+	scheduler.deleteTxn(startTS)
+	wakeupList := scheduler.latches.Release(txn.lock, commitTS)
+	for _, s := range wakeupList {
+		scheduler.run(s)
```
Comment (coocood) [inline store/tikv/txn_scheduler.go]: We should run wakeup transactions in background goroutine, but run directly at line 138

File: config/config.toml.example [hunk 13]
```diff
@@ -38,6 +38,9 @@ oom-action = "log"
 # Enable coprocessor streaming.
 enable-streaming = false
 
+# Enable local latches for transactions.
```
Comment (disksing) [inline config/config.toml.example]: Add description about under what circumstances should user consider turn the feature on.

File: store/tikv/txn_scheduler.go [hunk 14]
```diff
@@ -0,0 +1,189 @@
...
+}
+
+func (t *txnMap) put(startTS uint64, txn *txnCommitter) {
+	t.Lock()
+	defer t.Unlock()
+	if t.txns == nil {
+		t.txns = make(map[uint64]*txnCommitter)
+	}
+	if _, ok := t.txns[startTS]; ok {
+		panic(fmt.Sprintf("The startTS %d shouldn't be used in two transaction", txn.twoPC.startTS))
```
Comment (disksing) [inline store/tikv/txn_scheduler.go]: two transaction**s**.

File: store/tikv/txn_scheduler.go [hunk 15]
```diff
@@ -0,0 +1,189 @@
...
+func (scheduler *txnScheduler) execute(ctx context.Context, txn *twoPhaseCommitter, connID uint64) error {
+	newCommitter := newTxnCommiter(ctx, txn)
+	// latches disabled
+	if scheduler.latches == nil {
+		_, err := newCommitter.execute()
+		return errors.Trace(err)
+	}
+	// latches enabled
+	newCommitter.lock = scheduler.latches.GenLock(txn.startTS, txn.keys)
+	// for transactions which is not retryable, commit directly.
```
Comment (disksing) [inline store/tikv/txn_scheduler.go]: which **are** not retryable.

File: store/tikv/txn_scheduler.go [hunk 16]
```diff
@@ -0,0 +1,189 @@
...
+	if !sessionctx.GetRetryAble(ctx) {
+		err := scheduler.runForUnRetryAble(newCommitter)
+		return errors.Trace(err)
+	}
+	// for transactions which need to acquire latches
+	ch := make(chan error, 1)
+	newCommitter.ch = ch
```
Comment (disksing) [inline store/tikv/txn_scheduler.go]: Why not create `ch` when create the `committer`?

File: store/tikv/txn_scheduler.go [hunk 17]
```diff
@@ -0,0 +1,189 @@
...
+func (t *txnMap) delete(startTS uint64) {
+	t.Lock()
+	defer t.Unlock()
+	delete(t.txns, startTS)
+}
+
+// txnScheduler is the scheduler to commit transactions.
+type txnScheduler struct {
+	latches *latch.Latches
+	txnMaps []txnMap
```
Comment (disksing) [inline store/tikv/txn_scheduler.go]: Do we really need an array of `txnMap`? Maybe try try `sync.Map`?

From the document of `sync.Map`:
> The Map type is optimized for two common use cases: (1) when the entry for a given key is only ever written once but read many times, as in caches that only grow, or (2) when multiple goroutines read, write, and overwrite entries for disjoint sets of keys. In these two cases, use of a Map may significantly reduce lock contention compared to a Go map paired with a separate Mutex or RWMutex.

It looks like our scenario fits the description of case 2.

File: store/tikv/txn_scheduler.go [hunk 18]
```diff
@@ -0,0 +1,189 @@
...
+}
+
+const (
+	txnSchedulerLatchesSize = 102400
```
Comment (disksing) [inline store/tikv/txn_scheduler.go]: How much memory will it cost? Do we need to make it configurable?

File: store/tikv/txn_scheduler.go [hunk 19]
```diff
@@ -0,0 +1,189 @@
...
+	acquired, stale := scheduler.latches.Acquire(txn.lock)
+	if !stale && !acquired {
+		// wait for next wakeup
+		return
+	}
+	commitTS := txn.commit(stale)
+	scheduler.deleteTxn(startTS)
+	wakeupList := scheduler.latches.Release(txn.lock, commitTS)
+	for _, s := range wakeupList {
+		go scheduler.run(s)
```
Comment (disksing) [inline store/tikv/txn_scheduler.go]: Do we need to use the _goroutine pool_ here? cc @tiancaiamao

File: store/tikv/txn_scheduler.go (hunk 16 above)
Comment (AndreMouche) [reply store/tikv/txn_scheduler.go]: We do not `ch` when `latches` are disabled or transactions  are not retryable @disksing

File: store/tikv/latch/latch.go [hunk 20]
```diff
@@ -146,6 +157,13 @@ func (latches *Latches) Release(lock *Lock, commitTS uint64) (wakeupList []uint6
 	return
 }
 
+// RefreshCommitTS refresh slot's commitTS.
+func (latches *Latches) RefreshCommitTS(slotIDs []int, commitTS uint64) {
+	for _, slotID := range slotIDs {
+		latches.slots[slotID].refreshCommitTS(commitTS)
```
[the remark this answers is not in the document]
Comment (AndreMouche) [reply store/tikv/latch/latch.go]: It seems yes @tiancaiamao

**Comment by AndreMouche** [comment]
Do you mean we could just wait until acquiring all slot's lock? @tiancaiamao

**Pull request closed without merging**